src/core/net/srp_client.cpp

```diff
@@ -0,0 +1,1551 @@
+    // but not UDP/IP header before the header RR counts have been
+    // adjusted for the inclusion of SIG(0).
+
+    sha256.Start();
+
+    // (1) SIG RR RDATA wire format
+    len = aMessage.GetLength() - offset - sizeof(Dns::ResourceRecord);
+    sha256.Update(aMessage, offset + sizeof(Dns::ResourceRecord), len);
+
+    // (2) Message from DNS header before SIG
+    sha256.Update(aMessage, 0, offset);
+
+    sha256.Finish(hash);
+    SuccessOrExit(error = aInfo.mKeyPair.Sign(hash, signature));
+
+    // Move back in message and append SIG RR now with compressed host
+    // name (as signer's name) along with the calculated signature.
+
+    IgnoreError(aMessage.SetLength(offset));
+
+    // SIG(0) uses owner name of root (single zero byte).
+    SuccessOrExit(error = Dns::Name::AppendTerminator(aMessage));
+
+    offset = aMessage.GetLength();
+    SuccessOrExit(error = aMessage.Append(sig));
+    SuccessOrExit(error = AppendHostName(aMessage, aInfo));
+    SuccessOrExit(error = aMessage.Append(signature));
+    UpdateRecordLengthInMessage(sig, offset, aMessage);
+
+exit:
+    return error;
+}
+
+void Client::UpdateRecordLengthInMessage(Dns::ResourceRecord &aRecord, uint16_t aOffset, Message &aMessage) const
+{
+    // This method is used to calculate an RR DATA length and update
+    // (re-write) it in a message. This should be called immediately
+    // after all the fields in the record are written in the message.
+    // `aOffset` gives the offset in the message to the start of the
+    // record.
+
+    aRecord.SetLength(aMessage.GetLength() - aOffset - sizeof(Dns::ResourceRecord));
+    aMessage.Write(aOffset, aRecord);
+}
+
+void Client::HandleUdpReceive(void *aContext, otMessage *aMessage, const otMessageInfo *aMessageInfo)
+{
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    static_cast<Client *>(aContext)->ProcessResponse(*static_cast<Message *>(aMessage));
+}
+
+void Client::ProcessResponse(Message &aMessage)
+{
+    static const ItemState kNewStateOnUpdateDone[]{
+        /* (0) kToAdd      -> */ kToAdd,
+        /* (1) kAdding     -> */ kRegistered,
+        /* (2) kToRefresh  -> */ kToRefresh,
+        /* (3) kRefreshing -> */ kRegistered,
+        /* (4) kToRemove   -> */ kToRemove,
+        /* (5) kRemoving   -> */ kRemoved,
+        /* (6) kRegistered -> */ kRegistered,
+        /* (7) kRemoved    -> */ kRemoved,
+    };
+
+    otError             error = OT_ERROR_NONE;
+    Dns::UpdateHeader   header;
+    uint16_t            offset = aMessage.GetOffset();
+    uint16_t            recordCount;
+    LinkedList<Service> removedServices;
+
+    VerifyOrExit(GetState() == kStateUpdating);
+
+    SuccessOrExit(error = aMessage.Read(offset, header));
+
+    VerifyOrExit(header.GetType() == Dns::Header::kTypeResponse, error = OT_ERROR_PARSE);
+    VerifyOrExit(header.GetQueryType() == Dns::Header::kQueryTypeUpdate, error = OT_ERROR_PARSE);
+    VerifyOrExit(header.GetMessageId() == mUpdateMessageId, error = OT_ERROR_PARSE);
```

Comment:
nit: a different error to indicate message id mismatch?
```suggestion
    VerifyOrExit(header.GetMessageId() == mUpdateMessageId, error = OT_ERROR_DROP);
```